### PR TITLE
chore(consensus): add metrics for how long it takes to compute a bouncer function

### PR DESCRIPTION
**Pull request by kpop-dfinity**
Example:
Screenshot 2024-10-07 at 17 29 14

### CODE REVIEW

**Comment by eichhorl** [comment]
Thanks! Why is the average for IDKG and certifications increasing over time in your screenshot? I guess we also have HTTP outcalls and ingress bouncers, or do you expect them to not be very interesting?

**Comment by kpop-dfinity** [comment]
> Thanks! Why is the average for IDKG and certifications increasing over time in your screenshot?

To compute certification bouncer we call [CertificationPool::certified_heights](https://sourcegraph.com/github.com/dfinity/ic/-/blob/rs/artifact_pool/src/certification_pool.rs?L370-375) which iterates through all certifications, so over time the time needed to compute the bouncer should increase, until we purge the certifications. AFAIK we purge certifications only after a next CUP height is reached, which never happened in my test (the DKG interval is set to 999, so not enough time has elapsed)

For IDK, I'm not sure tbh. Maybe because of this line? https://sourcegraph.com/github.com/dfinity/ic/-/blob/rs/consensus/src/idkg.rs?L469


> I guess we also have HTTP outcalls and ingress bouncers, or do you expect them to not be very interesting?

I only needed the consensus one, to see if it's quick enough to safely decrease the refresh period, but since the other bouncers are defined in the same crate, I could easily add the metric for them as well, so I did. With http outcalls and ingress bouncers I would have to duplicate the code, which I didn't want to do (maybe you have a suggestion where would be a good place to share the code?) now to keep things simple. I'll add it in a later PR if necessary

**Comment by eichhorl** [comment]
Okay, I was confused by the 999 DKG interval, now it makes sense. I think the increase for IDKG may come from creating the `IDkgBlockReader` which needs to clone the growing blockchain cache. But I think the time still seems manageable

**Comment by eichhorl** [comment]
I'm fine with leaving out Ingress and HTTP outcalls, and only adding them if needed. Thanks for the explanation!

**Review by eichhorl — APPROVED** [review]

File: rs/consensus/src/bouncer_metrics.rs [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+impl BouncerMetrics {
+    pub(crate) fn new(registry: &MetricsRegistry, pool_type: &str) -> Self {
+        Self {
+            update_duration: registry.register(
+                Histogram::with_opts(histogram_opts!(
+                    "consensus_bouncer_update_duration",
+                    "How long it took to compute the bouncer function, in seconds",
+                    // 0.1ms, 0.2ms, 0.5ms, 1ms, 2ms, 5ms, 10ms, 20ms, 50ms, 100ms, 200ms, 500ms,
+                    // 1s, 2s, 5s, 10s, 20s, 50s
```
Comment (DSharifi) [inline rs/consensus/src/bouncer_metrics.rs:17]: Do we need buckets for the order of 10s of seconds? Seems like a big range to go from 0.1ms to 50s

File: rs/consensus/src/certification/certifier.rs [hunk 2]
```diff
@@ -75,6 +81,8 @@ impl<Pool: CertificationPool> BouncerFactory<CertificationMessageId, Pool> for C
     // any new artifacts at that height. If it is above the CUP height and we do not
     // have a full certification at that height, we're interested in all artifacts.
     fn new_bouncer(&self, certification_pool: &Pool) -> Bouncer<CertificationMessageId> {
+        let _timer = self.metrics.start_timer();
```
Comment (DSharifi) [inline rs/consensus/src/certification/certifier.rs:84]: I'm not sure this abstractions make it easy to follow what timer is being started. I would prefer if we just use the metric field directly, as it makes it less confusing what timer you are starting.
```suggestion
        let _timer = self.metrics.update_duration.start_timer();
```

**Review by DSharifi — APPROVED** [review]

File: rs/consensus/src/bouncer_metrics.rs (hunk 1 above)
Comment (kpop-dfinity) [reply rs/consensus/src/bouncer_metrics.rs:17]: Yeah, we don't, I've reduced the maximum to 500ms.

**Pull request merged**